VideoPlayer: Improve "no data" handling in our ffmpeg demuxer

a1rwulf opened the pull request:
## Description
Our ffmpeg demuxer has 2 issues:
* It freezes the UI for >20s when you read from streams that have no data
* It freezes the UI for ~30s when you start playback of a stream that has no data

This PR tries to improve both scenarios for livetv streams.
We do not want to freeze the UI for that big amount of time in either of the usecases.

## How Has This Been Tested?
Tested with live feeds and pvr.iptvsimple

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

@FernetMenta In case your're interested, your feedback is welcome (as usual) - thx.

basilgello commented:
I can incorporate it in the nightly build and check if it solves a 30-sec stall on channel switch with iptvsimple & timeshift.

FernetMenta commented:
@a1rwulf accessing inputstream and demuxer from other threads than VP's own thread is wrong (not thread safe). What you really want to fix is the silly 30sec timeout of CurlFile and the 10sec timeout of CFileCache. You don't want VP's thread to be blocked in Read() in case there is no data, which can happen during playback for many reasons (in particular on live tv). Read should never block for more than 1 sec or it results in bad user experience.

a1rwulf commented:
@FernetMenta the particular problem I face can probably be described best with following screenshot of the stacktrace:
![image](https://user-images.githubusercontent.com/5141257/98779488-3d28ab80-23f4-11eb-8cc6-1c9ad5a11f8c.png)

I don't see any relation to `CurlFile` or `CFileCache` in this case, while I still agree that those timeouts might be bad as well.

Demuxer and Inputstream are called from different threads already here:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L704
Does this need fixing too?

As an alternative solution:
Shall I try to make `OpenDemuxStream` non-blocking and move `Prepare` into the threadloop, so we can still handle incoming messages while the demuxer is setup?

a1rwulf commented:
> I can incorporate it in the nightly build and check if it solves a 30-sec stall on channel switch with iptvsimple & timeshift.

@basilgello can you point me to a report regarding that stall, I wonder if/how it's related?

FernetMenta commented:
> I don't see any relation to CurlFile or CFileCache in this case, while I still agree that those timeouts might be bad as well.

Pause the debugger when it hangs and you'll see that the VP thread is blocked in the Read() function of those classes. If VP's thread is blocked, VP can't update its state and won't accept new commands. The desired behaviour is that Read() doesn't block for long and the demuxer keeps going in case of no data. Consider a thunderstorm between dish and satellite. The expected behaviour is that video freezes and the user gets a visual cue (buffering control) that something happened. After the thunderstorm has gone away, video should continue.
Fix the blocking in Read() and it will solve your issue as well.

> Demuxer and Inputstream are called from different threads already here:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L704
Does this need fixing too?

This code is more than 10 years old and wrong. It can crash because it is not thread safe.

FernetMenta commented:
@a1rwulf from the stack trace I see that it does not use any of Kodi's inputstream classes but ffmpeg's internal read. What type of file is this?

a1rwulf commented:
@FernetMenta it's a liveTV channel played through pvr.iptvsimple which doesn't use it's own demuxer, so it goes through `CDVDDemuxFFmpeg`.
In my case it's a mix of iptv channels like `https://rbmn-live.akamaized.net/hls/live/2002825/geoSTVATweb/master_1628.m3u8` and udp multicasts.


> @a1rwulf from the stack trace I see that it does not use any of Kodi's inputstream classes but ffmpeg's internal read. What type of file is this?
> You should decrease this timeout: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L242

Yes that's the first timeout that's used when the ffmpeg demuxer is setup.
Then there also is:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L983

Which is the timeout used for the thunderstorm usecase you brought up.

Either way:
If I reduce one of those timeouts playback will fail immediately, so I try to figure out how to keep VP alive in those scenarios.

a1rwulf commented:
Alright, last commit changes the read timeout and doesn't Flush for `AVERROR_INVALIDDATA` or `AVERROR_EXIT` which results in the picture being frozen and a buffer spinner showing up.
User can still take any action like switch away from the channel or stop playback.

Now I need to take care what happens when I switch to a channel that doesn't give data from the beginning.
This is where VP's `Prepare` will freeze up, which is currently above the VP main loop.

FernetMenta commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -980,23 +980,22 @@ DemuxPacket* CDVDDemuxFFmpeg::Read()
       // timeout, probably no real error, return empty packet
       bReturnEmpty = true;
     }
-    else if (m_pkt.result == AVERROR_EOF)
```
When you change code you should always think of all use cases, not just your own you are trying to fix. If you drop this block, VP won't stop at the end of normal files. The code should like like this:

```
      m_timeout.Set(1000);
      m_pkt.result = av_read_frame(m_pFormatContext, &m_pkt.pkt);
      m_timeout.SetInfinite();
    }

    if (m_pkt.result == AVERROR(EINTR) || m_pkt.result == AVERROR(EAGAIN))
    {
      // timeout, probably no real error, return empty packet
      bReturnEmpty = true;
    }
    else if (m_pkt.result == AVERROR_EOF)
    {
      if (m_pInput->IsRealtime())
        bReturnEmpty = true;
    }
    else if (m_pkt.result < 0)
    {
      Flush();
      if (m_pInput->IsRealtime())
        bReturnEmpty = true;
    }
```

a1rwulf replied:
yeah sorry that was an accident - I didn't mean to remove the AVERROR_EOF code.
will push your version soon.

a1rwulf replied:
Just to clarify with the posted livestream I had issues on ts segment changes of the HLS stream when I used 100ms as timeout.
So I went with 500ms which seems to work fine, thought it's worth mentioning.

FernetMenta replied:
I won't set timeout to less than 1000ms. If there is an issue, it does not matter if you notice it after 500ms or 1000ms. But if you run into timeout without an actual issue, you'll get video corruption because ffmpeg does not like truncated reads.

a1rwulf replied:
Fair enough, thx for the info!

FernetMenta commented:
> Now I need to take care what happens when I switch to a channel that doesn't give data from the beginning.

the 30sec timeout in open is still bad. opening a mpegts stream doesn't take long, if done correctly. If that takes longer than 1 or 2 sec, you can assume that there is no data and can error out. that will result in an error message on screen, which is imo the best option for a non playable url.

a1rwulf committed:
VideoPlayer: DemuxFFmpeg - Improve errorhandling for livestreams

Instead of freezing for 20s and then abort playback,
whenever a LiveTV signal doesn't deliver any data we
want to just freeze the picture and continue playback
when we get data again.
At any time the user has the possibility to switch to
another channel or stop the playback.

a1rwulf commented:
> > Now I need to take care what happens when I switch to a channel that doesn't give data from the beginning.
> 
> the 30sec timeout in open is still bad. opening a mpegts stream doesn't take long, if done correctly. If that takes longer than 1 or 2 sec, you can assume that there is no data and can error out. that will result in an error message on screen, which is imo the best option for a non playable url.

So now, I'd like to tackle this.
My current observations:
If I reduce the timeout to 1000ms the demuxer setup fails quite often, so that's a nono.
A value of 3-5 seconds would be good I think, but then I still see 2 problems:
* When switching from a working channel to a broken one, the error dialog isn't shown, seems like a race condition
* When I switch from working, to not working, to working within those 3-5s I expect the new working channel to play, but instead VP stops with the error.

I will focus on the second issue for now, because imho that's way more annoying than the dialog not popping up.
Therefore I need a way to make VP not abort when a new playback gets scheduled while we're stuck in the previous demuxer creation.

a1rwulf commented:
I tried to outline my strategy in the commit message, but not sure if this is a shot in the right direction.

basilgello commented:
> > I can incorporate it in the nightly build and check if it solves a 30-sec stall on channel switch with iptvsimple & timeshift.
> 
> @basilgello can you point me to a report regarding that stall, I wonder if/how it's related?

I did not report it yet because it is extremely hard to log anything useful, but @phunkyfish is aware of it

FernetMenta commented:
> If I reduce the timeout to 1000ms the demuxer setup fails quite often, so that's a nono.

You missed one information I gave:

> opening a mpegts stream doesn't take long, if done correctly

If your stream does not open instantly, you are facing another issue you should try to solve. That would also improve user experience when switching channels. Most likely ffmpeg spends a lot of time in probing/tuning the mpegts stream. This can be disabled for realtime streams. Then opening the stream will be almost instant.

a1rwulf commented:
> If your stream does not open instantly, you are facing another issue you should try to solve. That would also improve user experience when switching channels. Most likely ffmpeg spends a lot of time in probing/tuning the mpegts stream. This can be disabled for realtime streams. Then opening the stream will be almost instant.

I'll try to get to the bottom of this.

a1rwulf committed:
VideoPlayer: Improve timoeut handling of demuxer creation

For ffmpeg demuxed livestreams a timeout of 30s really
hurts the user experience when broken channels are in
the game.
Keep 30s for non ffmpeg inputstreams, but use 5s for each
of the open calls for ffmpeg inputstreams.

a1rwulf commented:
Alright, I don't think the opening of streams is wrong.
Opening the channel I posted above takes ~2s.
Our old friend `avformat_find_stream_info` is not skipped in my case, but doesn't seem to be a problem either:
```
2020-11-12 08:57:43.162 T:3972    DEBUG <general>: Open - avformat_find_stream_info starting
2020-11-12 08:57:43.265 T:3972    DEBUG <general>: Open - av_find_stream_info finished
```

What I didn't get before was that the timeout is used for all the ffmpeg calls, as we call into ffmpeg multiple times if I set the timeout to 1s of course at some point it starts to fail.
So I need to set it as close to the ffmpeg calls as possible, which I do in the new commit.

After implementing the 3rd commit I found that 5s is a good timeout, because it allows you to browse between channels without getting kicked into the menu with an error.

Hope my explanation makes sense.

FernetMenta commented:
> Alright, I don't think the opening of streams is wrong.

Maybe not wrong, but not ideal. mpeg-ts should open instantly.

> Our old friend avformat_find_stream_info

This is not the problem anymore after you upgraded ffmpeg. Now the time is wasted in avformat_open_input.

a1rwulf commented:
> This is not the problem anymore after you upgraded ffmpeg. Now the time is wasted in avformat_open_input.

avformat_open_input takes around 300ms, we do it twice (don't know the reason) so this takes around 600ms in total I suppose.
Channelswitch-Time for my test was around 1.5s - 2s which I consider normal to be honest.

Anyway, any thoughts about the code I wrote for not aborting VP when you do a channelup within the set timeout?

FernetMenta commented:
> Anyway, any thoughts about the code I wrote for not aborting VP when you do a channelup within the set timeout?

I don't see how this changes behaviour. The ABORT message you send should be handled by VP right after is was sent. From an architecture POV it is quite ugly because the the sender and receiver is the same thread.

FernetMenta commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -239,9 +239,6 @@ bool CDVDDemuxFFmpeg::Open(const std::shared_ptr<CDVDInputStream>& pInput, bool
   m_pFormatContext->interrupt_callback = int_cb;
 
-  // try to abort after 30 seconds
-  m_timeout.Set(30000);
```
I would change the timeout only for realtime streams. Donˋt forget the users who want to play rar files from a raspberry pi over some slow network connection.

a1rwulf replied:
Ah yeah nice idea, will do.

a1rwulf replied:
I changed the timeout of first codepath to 15s per call, which equals the 30sec as we do 2 calls.
For mms files it's 15s because only 1 call is done if I read it correctly.
I hope that's rather OK at least I'd already press stop if start of playback takes more than 10sec.
If there are objections I can raise it to whatever value we are comfortable with.

FernetMenta replied:
You change the timeout from 30 to 15 in beta stage. In this case you should give a reason why you think elupus was wrong setting it to 30 https://github.com/xbmc/xbmc/commit/9cd4a305f24830fc14582e08307bb58da1f4200b

a1rwulf replied:
Fixed, I'll keep the 30s.

a1rwulf commented:
> > Anyway, any thoughts about the code I wrote for not aborting VP when you do a channelup within the set timeout?
> 
> I don't see how this changes behaviour. The ABORT message you send should be handled by VP right after is was sent. From an architecture POV it is quite ugly because the the sender and receiver is the same thread.

I thought it would make sense to use the existing message loop.
Can you explain what exactly makes using the message queue from the same thread ugly in terms of architecture?
Do you see any other good option to do this?

> 
> EDIT: now I got it. The new open cmd is sent while VP processes Prepare(). In this case the open msg will be processed before the abort msg. Makes sense to me.

Yes, exactly that was my idea, the open message has priority and when a fresh open is successfully prepared I clear the pending abort.

a1rwulf commented:
@FernetMenta sorry I don't mean to be pushy, but I'd like to get it into v19 if possible as I think it's a good improvement.
Do you consider the current state good enough to be pushed, or is the architectural concern you mentioned big enough that I should reconsider the implementation?

FernetMenta commented:
> Do you consider the current state good enough to be pushed

Wrong question. I don't care anymore what you push. I can give you my judgement on the proposed change in terms of good or bad.
As already mentioned it is an abuse of the msg queue design pattern and as a result hard to understand. Even I struggled on first glance how this was supposed to work. Further this change does not cover all cases because of a race condition. At the time the user selects the new item you don't have guarantee that VP would accept it because it might already handle the abort message.

If players weren't singletons, the solution would be easy. Just create a new player instance and abandon the old one. At the time I was active I did lots of small steps to finally reach that goal. There isn't much left that prevents VP from having multiple instances.

a1rwulf commented:
I'm not here to add race conditions or violate patterns.

> If players weren't singletons, the solution would be easy. Just create a new player instance and abandon the old one. At the time I was active I did lots of small steps to finally reach that goal. There isn't much left that prevents VP from having multiple instances.

I'll drop the cancellation for now and see if I can get above for v20 instead.

bantilan reviewed (changes requested):
This cause an error playing some live stream content:

2021-02-02 17:06:55.019 T:716      INFO <general>: VideoPlayer::OpenFile: pvr://channels/tv/All%20channels/pvr.iptvsimple_182077386.pvr
2021-02-02 17:06:55.020 T:1392     INFO <general>: Creating InputStream
2021-02-02 17:06:55.340 T:1392     INFO <general>: Creating Demuxer
2021-02-02 17:07:01.572 T:1392    ERROR <general>: OpenDemuxStream - Error creating demuxer
2021-02-02 17:07:01.572 T:1392     INFO <general>: CVideoPlayer::OnExit()
2021-02-02 17:07:01.715 T:716      INFO <general>: CVideoPlayer::CloseFile()
2021-02-02 17:07:01.716 T:716      INFO <general>: VideoPlayer: waiting for threads to exit
2021-02-02 17:07:01.716 T:716      INFO <general>: VideoPlayer: finished waiting

Example video codec: 
CODECS="avc1.640032,mp4a.40.2",RESOLUTION=1920x1080,FRAME-RATE=30.000

lrusak commented:
> This cause an error playing some live stream content:
> 

@bantilan this PR hasn't even been merged. Are you applying it to your build?

bantilan commented:
> > This cause an error playing some live stream content:
> 
> @bantilan this PR hasn't even been merged. Are you applying it to your build?

yes on my fork version.. i am using LibreELEC.

a1rwulf commented:
That's actually interesting, as I wouldn't think that it has side effects like this.
But as it's WIP and isn't going to be merged in it's current state, I don't really want to bother.
